### PR TITLE
metrics: Send samples per series instead of per metric name

### PR DESCRIPTION
**Pull request by d3dave**
Currently samples are only generated once per metric name, ignoring additional labels which might result in additional series. This PR standardizes the data classes to be used for samples and the entire snapshot, and fixes the code to report samples for all series even if they have the same metric name but different labels.

### CODE REVIEW

File: gprofiler/spark/spark_collector.py [hunk 1]
```diff
@@ -806,15 +754,18 @@ def _collect_loop(self) -> None:
                         break
 
             if self._spark_sampler is not None:
-                metrics = list(self._spark_sampler.collect())
-                timestamp = datetime.now(tz=timezone.utc)
+                collected = self._spark_sampler.collect()
+                # No need to submit samples that don't actually have a value:
```
Comment (Jongy) [inline gprofiler/spark/spark_collector.py:758]: Where does it actually happen? Or when?

File: gprofiler/spark/spark_collector.py [hunk 2]
```diff
@@ -806,15 +754,18 @@ def _collect_loop(self) -> None:
...
+                samples = tuple(filter(lambda s: s.value is not None, collected))
+                snapshot = MetricsSnapshot(datetime.now(tz=timezone.utc), samples)
+                # _client and _storage_dir are mutually exclusive, so we will never bake the payload twice
```
Comment (Jongy) [inline gprofiler/spark/spark_collector.py:761]: Are they? You can pass both.

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing

File: gprofiler/spark/spark_collector.py (hunk 2 above)
Comment (d3dave) [reply gprofiler/spark/spark_collector.py:761]: `_client` => `--upload-results`, and `_storage_dir` => `--output-dir`, and the two flags are mutually exclusive, so de facto so are the fields. No?

File: gprofiler/spark/spark_collector.py (hunk 1 above)
Comment (d3dave) [reply gprofiler/spark/spark_collector.py:758]: If a json field specified in one of the `METRICS` dicts is not present in a json response returned from the yarn/spark server, then a sample can be reported for that metric with value `None`.

Comment (Jongy) [reply gprofiler/spark/spark_collector.py:758]: I see. We have logs for that now though, so we'll need to add if we ever encounter..

File: gprofiler/spark/spark_collector.py (hunk 2 above)
Comment (Jongy) [reply gprofiler/spark/spark_collector.py:761]: They are not mutual exclusive o.O

**Review by Jongy — APPROVED** [review]

**Pull request merged**